### ISSUE TITLE
Slow merging process and memory limit exceeded

### ISSUE THREAD

**Issue by tanchik-curly** [ClickHouse/ClickHouse#67972]
Hi, do you know why my merging takes more than 10 minutes (usually up to 1 hour), I tried to configure 
```
background_pool_size=36
background_fetches_pool_size = 18
```
, I tried to set bigger values, but nothing, I have 72 cores per replica, 148Gb of RAM.

I usually get this message in postpone_reason:

`(MergeFromLogEntryTask): virtual bool DB::ReplicatedMergeMutateTaskBase::executeStep(): Code: 241. DB::Exception: Memory limit (total) exceeded: would use 113.34 GiB (attempt to allocate a chunk of 4196320 bytes), maximum: 109.80 Gb. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero.: (while reading column xxx): located on disk default of type local, from mark 2 with max_rows_to_read = 4617): While executing MergeTreeSequentialSource. (MEMORY_LIMIT_EXCEEDED)`

I do optimize table <> final deduplicate quite often to avoid having duplicates
Speaking about other errors, I also usually get:

`DB::StorageReplicatedMergeTree::processQueueEntry(ReplicatedMergeTreeQueue::SelectedEntryPtr)::(anonymous class)::operator()(LogEntryPtr &) const: Code: 243. DB::Exception: Cannot reserve 1.00 MiB, not enough space. (NOT_ENOUGH_SPACE)`

could it be a problem with background_fetches_pool_size?

![image](https://github.com/user-attachments/assets/4690d1da-f23b-4268-b2f8-313535dbd66a)

**Comment by den-crane** [comment]
Please provide Clickhouse version.
Do you have very long strings in this table? Or do you have something like uniqExact aggregate function? Provide `show create table ...`

**Comment by den-crane** [comment]
I think I misunderstood the initial question. 
So everything works fine but you observe some errors and retries?
Can you please share information about some of these long merges: `select * from system.part_log where elapsed > .... format Vertical limit 5`.

**Comment by tanchik-curly** [comment]
Yes, exactly, I have a problem only with the merging process, it takes too much time to merge tables (up to 1 hour), I need to duplicate parts manually, I started to receive more error logs, like: 

`(MergeFromLogEntryTask): virtual bool DB::ReplicatedMergeMutateTaskBase::executeStep(): Code: 241. DB::Exception: Memory limit (total) exceeded: would use 113.34 GiB (attempt to allocate a chunk of 4196320 bytes), maximum: 109.80 Gb. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero.: (while reading column xxx): located on disk default of type local, from mark 2 with max_rows_to_read = 4617): While executing MergeTreeSequentialSource. (MEMORY_LIMIT_EXCEEDED)`

Attached the files with merge table results and part_log table for involved data parts, generated at 07.30 UTC

[merge_table.csv](https://github.com/user-attachments/files/16539362/merge_table.csv)
[part_table.csv](https://github.com/user-attachments/files/16539364/part_table.csv)

**Comment by azat** [comment]
>do you think it could be caused by full table optimization and a big task pool?

Yes, definitely. The bigger the pool the more merges you can have, and you need more memory and disk. I would recommend you avoid having such huge merges.

I looked at the `system.merges` that you provided, for instance result part `ebdd9fdf60d3709c23ed11242260ddd2_0_185008_3410_452`, and it writes 500GiB of data, and executed for almost half an hour, so it is 266MiB/s, and this is only write speed, likely you have SSDs, but I am not that in one thread w/o async writes it is possible write faster to disk.

So I would say that everything is expected here.

Feel free to reopen the issue if you will have more questions.
